docs/README.md

```diff
@@ -0,0 +1,256 @@
+
+```
+set AZURE_TENANT_ID=<YOUR_TENANT_ID>
+```
+
+In Unix based systems, you could use the command such as:
+
+```
+export AZURE_TENANT_ID=<YOUR_TENANT_ID>
+```
+
+### Option 2 - Options Hash
+The initialization of profile clients take an options hash as a parameter. This options hash consists of tenant_id, client_id, client_secret, subscription_id, active_directory_settings and credentials. Among these, the active_directory_settings and credentials are optional.
+
+You can set the (above) values using the options hash:
+
+```ruby
+options = {
+  tenant_id: 'YOUR TENANT ID',
+  client_id: 'YOUR CLIENT ID',
+  client_secret: 'YOUR CLIENT SECRET',
+  subscription_id: 'YOUR SUBSCRIPTION ID'
+}
+```
+
+## OAuth2 Token Credentials
+
+[Acquire an acces token](https://docs.microsoft.com/en-us/rest/api/azure/#acquire-an-access-token)
+
+### Use token provider
+
+There are 4 kinds of token providers in library:
+
+* MsRest::StringTokenProvider
+* MsRestAzure::ApplicationTokenProvider
+* MsRestAzure::AzureCliTokenProvider
+* MsRestAzure::MSITokenProvider
+
+In most cases, we use the ApplicationTokenProvider. You could use the following code
+
+```ruby
+provider = MsRestAzure::ApplicationTokenProvider.new(
+       'YOUR TENANT ID',
+       'YOUR CLIENT ID',
+       'YOUR CLIENT SECRET')
+credentials = MsRest::TokenCredentials.new(provider)
+```
+
+MSITokenProvider supports 2 ways to get MSI(Managed Service Identity) Token
+
+* Accessing the http://169.254.169.254/metadata/identity/oauth2/token to get the MSI Token (default)
+* Using the extension installed locally and accessing http://localhost:50342/oauth2/token to get the MSI Token
+
+You could use the following code
+
+```ruby
+provider = MsRestAzure::MSITokenProvider.new()
+credentials = MsRest::TokenCredentials.new(provider)
+```
+
+AzureCliTokenProvider use [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/) to acquire token. Azure CLI should be installed in PATH. You could use the following code
+
+```ruby
+provider = MsRestAzure::AzureCliTokenProvider.new()
+credentials = MsRest::TokenCredentials.new(provider)
+```
+
+### ActiveDirectoryServiceSettings
+
+ActiveDirectoryServiceSettings class represents a settings for Azure AD authentication. It contains 4 sets of properties for different clouds
+
+* get_azure_settings
+* get_azure_china_settings
+* get_azure_german_settings
+* get_azure_us_government_settings
+
+So you can create a application token provider with Azure China Settings by using the following code
+
+```ruby
+settings = MsRestAzure::ActiveDirectoryServiceSettings.get_azure_china_settings
+provider = MsRestAzure::ApplicationTokenProvider.new(
+       'YOUR TENANT ID',
+       'YOUR CLIENT ID',
+       'YOUR CLIENT SECRET',
+       settings)
+credentials = MsRest::TokenCredentials.new(provider)
+```
+
+## Make an async request 
+
+AzureServiceClient can be used to make an async request. You could use it directly or create a class that inherits it.
```

Comment:
is async request the standard pattern in Ruby? if so, we should provide an async request code sample

Reply:
All the samples will be asynchronized.